cohortextractor/query_language.py

```diff
@@ -19,8 +16,47 @@ def table(name):
     return Table(name)
 
 
-def category_group(source, comparison, value):
-    return CategoryGroup(source, _OPERATOR_MAPPING[comparison], value)
+class Comparator:
+    """A generic comparator to represent a comparison between a source object and a value"""
+
+    def __init__(
+        self, children=None, connector="and_", source=None, operator=None, value=None
+    ):
+        """
+        Construct a new Comparator.
+        A single comparator will have a source, operator and value.  A tree of Compararors
+        will have at most two child Comparators, which are to be connected with self.connector.
+        Each child may itself have more child Comparators, again with a connector to indicate
+        how they should be joined
+        """
```

Comment:
I feel there are multiple types trying to get out here. `Predicate`, a generic tree type, and `Comparator` and  `Conjunction` (or something) the concrete types that are the nodes and leaves in the tree. `connector` and `children` are fields only on `Conjunction`, whereas `source`, `operator` and `value` are fields only on `Comparator` (I think).

Reply:
I think these could be two things; so a comparison on a Value returns a Comparator, and a combination of Comparators returns a Conjunction which has children, which can be either Comparators or Conjunctions.

Reply:
Yes, agreed.

Reply:
Do we need all those types? I was thinking we've basically got an Expression object which has a type and then one or more arguments.

Possibly it's useful to have a BooleanExpression subclass whose type is always boolean regardless of the types of its arguments (which would cover all the comparison cases).